src/order/extension/well.lean

```diff
@@ -0,0 +1,66 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Junyan Xu
+-/
+import data.prod.lex
+import set_theory.ordinal.arithmetic
+
+/-!
+# Extend a well-founded order to a well-order
+
+This file constructs a well-order (linear well-founded order) which is an extension of a given
+well-founded order.
+
+## Proof idea
+
+We can map our order into two well-orders:
+* the first map respects the order but isn't necessarily injective. Namely, this is the *rank*
+  function `rank : α → ordinal`.
+* the second map is injective but doesn't necessarily respect the order. This is an arbitrary
+  well-order on `α`.
+
+Then their lexicographic product is a well-founded linear order which our original order injects in.
+-/
+
+universe u
+
+variables {α : Type u} {r : α → α → Prop}
+
+namespace well_founded
+variable (hwf : well_founded r)
+include hwf
+
+/-- An arbitrary well order on `α` that extends `r`. -/
```

Comment:
```suggestion
/-- An arbitrary well order on `α` that extends `r`.

The construction maps `r` into two well-orders, the first map is not injective but respects the
order, the other map is injective but doesn't respect the order. By taking the lexicographic product
of the two, we get both properties, so we can pull it back and extend our order into a well-order.

Another way to view this is that we choose an arbitrary well-order to serve as a tiebreak between
two unorderable elements.
-/
```

Reply:
Notice that Yaël already added a module docstring explaining the proof, which IMO should be removed if we explain it here. I suggest the following version that merges both:
```suggestion
The construction maps `r` into two well-orders: the first map is `well_founded.rank`, which is not
necessarily injective but respects the order `r`; the other map is the identity (with an arbitrarily
chosen well-order on `α`), which is injective but doesn't respect `r`. By taking the lexicographic
product of the two, we get both properties, so we can pull it back and get an well-order that extend
our original order `r`. Another way to view this is that we choose an arbitrary well-order to serve
as a tiebreak between two elements of same rank.
```

Reply:
I prefer to keep the proof idea in the docstring because `well_founded.well_order_extension` isn't the only result in this file that uses this idea.